### PR TITLE
DOPE-258: added positionalParameters to DopeQuery and added it to each toDopeQuery() and test

### PR DESCRIPTION
**Pull request by jansigi**
I am very sorry if you are reviewing this

### CODE REVIEW

**Review by martinagallati-ergon — CHANGES_REQUESTED** [review]
Overall great!
I do not really see the benefit in testing every expression with (almost) every combination of parameters; what do these tests really _test_, the function/expression itself or parameters? In my opinion a few build tests with parameters of different types would benefit us more, what do you think?
If we leave them in, I would request changing all the literal assignments in the tests to the helper functions.

File: core/src/main/kotlin/ch/ergon/dope/DopeQuery.kt [hunk 1]
```diff
@@ -1,3 +1,3 @@
 package ch.ergon.dope
 
-data class DopeQuery(val queryString: String, val parameters: Map<String, Any>)
+data class DopeQuery(val queryString: String, val parameters: Map<String, Any>, val positionalParameters: List<Any>)
```
[the remark this answers is not in the document]
Comment (pgruntz) [reply core/src/main/kotlin/ch/ergon/dope/DopeQuery.kt:3]: agree 👍

File: core/src/main/kotlin/ch/ergon/dope/resolvable/fromable/UseIndex.kt [hunk 2]
```diff
@@ -20,7 +20,11 @@ class IndexReference(
     private val indexName: String? = null,
     private val indexType: IndexType? = null,
 ) : Resolvable {
-    override fun toDopeQuery(manager: DopeQueryManager) = DopeQuery(formatIndexToQueryString(indexName, indexType?.type), emptyMap())
+    override fun toDopeQuery(manager: DopeQueryManager) = DopeQuery(
+        queryString = formatIndexToQueryString(indexName, indexType?.type),
+        parameters = emptyMap(),
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/fromable/UseIndex.kt:25]: can we maybe make these _emptyMaps_ and _emptyList_ as default values?

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/FromClause.kt [hunk 3]
```diff
@@ -18,7 +18,8 @@ class FromClause(private val fromable: Fromable, private val parentClause: ISele
         }
         return DopeQuery(
             queryString = formatToQueryStringWithSymbol(parentDopeQuery.queryString, "FROM", fromableDopeQuery.queryString),
-            parameters = fromableDopeQuery.parameters + parentDopeQuery.parameters,
+            parameters = parentDopeQuery.parameters + fromableDopeQuery.parameters,
+            positionalParameters = parentDopeQuery.positionalParameters + fromableDopeQuery.positionalParameters,
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/FromClause.kt:22]: Can we merge positionalParamters like that? Let's assume parentDopeQuery and fromableDopeQuery have both two positionalParameters. If we merge it like that, then parentDopeQuery.queryString has $1 and $2 and fromableDopeQuery.queryString has $3 and $4 is this always guaranteed?

**Review by pgruntz — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/ch/ergon/dope/DopeQuery.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/DopeQuery.kt]: I renamed the parameters to `namedParameters`.

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/GroupByClause.kt [hunk 4]
```diff
@@ -27,6 +27,9 @@ class GroupByClause(
             parameters = fieldsDopeQuery.fold(fieldDopeQuery.parameters) { fieldParameters, field ->
                 fieldParameters + field.parameters
             } + parentDopeQuery.parameters,
+            positionalParameters = parentDopeQuery.positionalParameters + fieldsDopeQuery.fold(
+                fieldDopeQuery.positionalParameters,
+            ) { fieldParameters, field -> fieldParameters + field.positionalParameters },
```
[the remark this answers is not in the document]
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/GroupByClause.kt]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/ReturningClause.kt [hunk 5]
```diff
@@ -30,6 +30,9 @@ sealed class ReturningClause(
             parameters = fieldsDopeQuery.fold(fieldDopeQuery.parameters) { fieldParameters, field ->
                 fieldParameters + field.parameters
             } + parentDopeQuery.parameters,
+            positionalParameters = parentDopeQuery.positionalParameters + fieldsDopeQuery.fold(
+                fieldDopeQuery.positionalParameters,
+            ) { fieldParameters, field -> fieldParameters + field.positionalParameters },
```
[the remark this answers is not in the document]
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/ReturningClause.kt]: done, now the parent always calls the new `merge()` function.

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/OrderByClause.kt [hunk 6]
```diff
@@ -40,6 +41,7 @@ class SelectOrderByTypeClause(
         return DopeQuery(
             queryString = formatToQueryStringWithSymbol(parentDopeQuery.queryString, ORDER_BY, stringDopeQuery.queryString + " $orderByType"),
             parameters = stringDopeQuery.parameters + parentDopeQuery.parameters,
+            positionalParameters = stringDopeQuery.positionalParameters + parentDopeQuery.positionalParameters,
```
[the remark this answers is not in the document]
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/OrderByClause.kt]: done

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Primitive.kt [hunk 7]
```diff
@@ -37,30 +38,33 @@ class StringPrimitive(value: String) : Primitive<StringType>(
...
-    {
-            manager: DopeQueryManager ->
+    { manager: DopeQueryManager ->
         collection.map { it.toDopeQuery(manager) }.let { dopeQueries ->
             DopeQuery(
                 queryString = formatListToQueryStringWithBrackets(dopeQueries, prefix = "[", postfix = "]"),
                 parameters = dopeQueries.fold(emptyMap()) { parameters, dopeQueryElement -> parameters + dopeQueryElement.parameters },
+                positionalParameters = dopeQueries.fold(emptyList()) { positionalParameters, dopeQueryElement ->
+                    positionalParameters + dopeQueryElement.positionalParameters
+                },
```
[the remark this answers is not in the document]
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/Primitive.kt]: obsolete

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/function/stringfunction/ConcatExpression.kt [hunk 8]
```diff
@@ -26,6 +26,10 @@ class ConcatExpression(
             parameters = firstStringDopeQuery.parameters + secondStringDopeQuery.parameters + stringTypesDopeQuery.fold(
                 emptyMap(),
             ) { stringTypeParameters, field -> stringTypeParameters + field.parameters },
+            positionalParameters = firstStringDopeQuery.positionalParameters + secondStringDopeQuery.positionalParameters +
+                stringTypesDopeQuery.fold(emptyList()) { stringTypePositionalParameters, field ->
+                    stringTypePositionalParameters + field.positionalParameters
+                },
```
[the remark this answers is not in the document]
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/function/stringfunction/ConcatExpression.kt]: obsolete

File: core/src/main/kotlin/ch/ergon/dope/resolvable/fromable/UseIndex.kt (hunk 2 above)
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/fromable/UseIndex.kt]: done, `DopeQuery` can now be initialized with only the `queryString` and the parameters will be empty.

File: core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/FromClause.kt (hunk 3 above)
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/clause/model/FromClause.kt]: With the new merge functionality the order is guaranteed as long as the parent DopeQuery calls the merge function.

**Comment by martinagallati-ergon** [comment]
Main changes are in file `DopeQuery` and I added tests in `ParameterizedTest` to check the order of parameters, otherwise it's mainly renaming and the addition of the new data class in each file.

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/aggregator/AggregateExpression.kt [hunk 9]
```diff
@@ -8,9 +8,9 @@ import ch.ergon.dope.resolvable.operator.FunctionOperator
 import ch.ergon.dope.validtype.ValidType
 
 sealed class AggregateExpression<T : ValidType>(
+    private val symbol: String,
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/aggregator/AggregateExpression.kt:11]: nice 👍

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/conditional/CaseExpression.kt [hunk 10]
```diff
@@ -39,9 +40,9 @@ class CaseExpression<T : ValidType, U : ValidType>(
                 conditionDopeQueries.joinToString(separator = " ", prefix = " ", postfix = " ") {
                     "$WHEN ${it.first.queryString} $THEN ${it.second.queryString}"
                 } + END,
-            parameters = caseDopeQuery.parameters + conditionDopeQueries.fold(emptyMap()) { parameters, query ->
-                parameters + query.first.parameters + query.second.parameters
-            },
+            parameters = caseDopeQuery.parameters.merge(
+                *conditionDopeQueries.map { it.first.parameters; it.second.parameters }.toTypedArray(),
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/conditional/CaseExpression.kt:44]: this is not correct, the params of first will get ignored: 
```
conditionDopeQueries.map {
    it.first.parameters; //This is just a statement and does nothing
    it.second.parameters //This is the return value of the map-function
}
```

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/conditional/CaseExpression.kt [hunk 11]
```diff
@@ -64,9 +65,9 @@ class ElseCaseExpression<T : ValidType, U : ValidType>(
                     "$WHEN ${it.first.queryString} $THEN ${it.second.queryString}"
                 } + "$ELSE ${elseCaseDopeQuery.queryString} " +
                 END,
-            parameters = caseDopeQuery.parameters + conditionDopeQueries.fold(emptyMap()) { parameters, query ->
-                parameters + query.first.parameters + query.second.parameters
-            } + elseCaseDopeQuery.parameters,
+            parameters = caseDopeQuery.parameters.merge(
+                *conditionDopeQueries.map { it.first.parameters; it.second.parameters }.toTypedArray(),
```
Comment (pgruntz) [inline core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/conditional/CaseExpression.kt:69]: same comment as above

**Review by pgruntz — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/aggregator/AggregateExpression.kt (hunk 9 above)
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/aggregator/AggregateExpression.kt:11]: 😎

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/conditional/CaseExpression.kt (hunk 10 above)
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/conditional/CaseExpression.kt]: done, added correct merging and a test

File: core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/conditional/CaseExpression.kt (hunk 11 above)
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/resolvable/expression/unaliased/type/conditional/CaseExpression.kt]: done

File: core/src/main/kotlin/ch/ergon/dope/DopeQuery.kt [hunk 12]
```diff
@@ -1,3 +1,29 @@
 package ch.ergon.dope
 
-data class DopeQuery(val queryString: String, val parameters: Map<String, Any>)
+data class DopeQuery(
+    val queryString: String,
+    val parameters: DopeParameters = DopeParameters(),
+)
+
+data class DopeParameters(
```
[the remark this answers is not in the document]
Comment (pgruntz) [reply core/src/main/kotlin/ch/ergon/dope/DopeQuery.kt:8]: And can you maybe add a small KDoc / Comment above the data class explaining shortly what namedParameters and positionalParameters are. And also mentioning that the positionalParameters are a sorted list of the values

**Review by pgruntz — APPROVED** [review]

File: core/src/main/kotlin/ch/ergon/dope/DopeQuery.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (martinagallati-ergon) [reply core/src/main/kotlin/ch/ergon/dope/DopeQuery.kt]: done

**Review by martinagallati-ergon — APPROVED** [review]

**Pull request merged**